### PR TITLE
Converging and running commands on a running container

### PR DESCRIPTION
**Pull request by mwrock**
This PR adds the ability to interact with an already running container without destroying and recreating it. If the driver finds that the container is already running with the command in its docker options, it will use the docker API's `exec` to run commands in the container. This implement's the driver function `connect_to_machine` so that one can call `machine_execute` and other commands that do not go through `ready_machine`.

This also persists the docker options in the node. It does this so that calls to `machine_execute` do not need to re inject `machine_options`.

I also added a few other small things:
- added standard items to .gitignore
- removed the Gemfile.lock
- changed the gemspec to work with the latest chef-provisioning gems
- monkey patched the `EventsOutputStream` to get stdout working on converge

I do think this could use more validation and which I will do but wanted to add this now to get another pair of eyes looking at it.

### CODE REVIEW

File: lib/chef/provisioning/docker_driver/driver.rb [hunk 1]
```diff
@@ -103,8 +112,8 @@ def build_container(machine_spec, machine_options)
         "#{source_repository}:#{source_tag}"
       else
         target_repository = 'chef'
-        target_tag = machine_spec.name
-
+        target_tag = machine_spec.location['container_name'] || machine_spec.name
```
Comment (marc-) [inline lib/chef/provisioning/docker_driver/driver.rb]: Attribute `location` seems obsolete. I would suggest to use `reference` instead.

**Comment by marc-** [comment]
From my perspective at first place `docker_transport` needs to be rewritten. So instead of run container on every command it needs to run long living process inside container and then use docker `execute` command for converge and running intermediate commands.

**Comment by mwrock** [comment]
@marc- thanks for taking a look! That was my initial thought as well. However there is a chicken and egg problem with starting a container with the "long lived" command and THEN doing converge. Typically, the converge must happen before the long lived command is available. And once a container is started via `start` or `run` you cannot change that command. So the first convergence strategy currently implemented in `docker-transport` where you build on the container image seems to be the right approach.

That said, it would be nicer if it was possible to kick of the running container and stick with that so if there is a more elegant approach to this that would be great. I just was not able to find one.

**Comment by marc-** [comment]
Long living command could be just `/bin/sh -c 'while true;do true; done'` or something. Once converge done, container gets committed (as how it happens now) and new image will be run with specified command. What do you think?

**Comment by mwrock** [comment]
Ahh gotcha. So you'd still create a new container but be able to leverage exec throughout the initial converge. That does sound better.

**Comment by marc-** [comment]
BTW, did you try to run example code?
```ruby
require 'chef/provisioning/docker_driver'

machine_image 'web_server' do
  recipe 'apache'

  machine_options :docker_options => {
      :base_image => {
          :name => 'ubuntu',
          :repository => 'ubuntu',
          :tag => '14.04'
      }
  }
end

machine 'web00' do
  from_image 'web_server'

  machine_options :docker_options => {
      :command => '/usr/sbin/httpd'
  }
end
```
I think your code is not 100% compatible with chef-provisioning 1.x . Please see #37 and #47 .

**Comment by mwrock** [comment]
I was using a different recipe. But it was not using machine_image which is where I'm guessing that the problem lies. I'll give this a go with this example.

**Comment by mwrock** [comment]
@marc-  Getting back to this...

Yup. the example fails because my PR does not address the from_image/machine_image issues. I think thats out of scope for this PR and seems to be addressed in your PR which should be accepted on its own to attend to that issue.

Does that sound fair?

**Comment by marc-** [comment]
Sounds good to me.

**Comment by mwrock** [comment]
FYI: I made more headway here based on @marc- 's comments and will be submiting this as a different PR since it is a more deeper refactoring.

**Pull request closed without merging**